src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java

```diff
@@ -130,79 +136,132 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
-            readVoidPairs = rddReadPairs; //no sort
-        } else if (numReducers > 0) {
-            readVoidPairs = rddReadPairs.sortByKey(comparator, true, numReducers);
+        if ( comparator == null ) {
+            return reads;
         } else {
-            readVoidPairs = rddReadPairs.sortByKey(comparator);
+            return sort(reads, comparator, numReducers);
         }
-        return readVoidPairs.keys();
     }
 
     //Returns the comparator to use or null if no sorting is required.
     private static Comparator<SAMRecord> getSAMRecordComparator(final SAMFileHeader header) {
         switch (header.getSortOrder()){
             case coordinate: return new HeaderlessSAMRecordCoordinateComparator(header);
-            case duplicate:
+            //duplicate isn't supported because it doesn't work right on headerless SAMRecords
+            case duplicate: throw new UserException.UnimplementedFeature("The sort order \"duplicate\" is not supported in Spark.");
             case queryname:
             case unsorted:   return header.getSortOrder().getComparatorInstance();
             default:         return null; //NOTE: javac warns if you have this (useless) default BUT it errors out if you remove this default.
         }
     }
+
+    /**
+     *   do a total sort of an RDD so that all the elements in partition i are less than those in partition i+1 according to the given comparator
+     */
+    private static <T> JavaRDD<T> sort(JavaRDD<T> reads, Comparator<T> comparator, int numReducers) {
+        Utils.nonNull(comparator);
+        Utils.nonNull(reads);
+
+        // Turn into key-value pairs so we can sort (by key). Values are null so there is no overhead in the amount
+        // of data going through the shuffle.
+        final JavaPairRDD<T, Void> rddReadPairs = reads.mapToPair(read -> new Tuple2<>(read, (Void) null));
+
+        final JavaPairRDD<T, Void> readVoidPairs;
+        if (numReducers > 0) {
+            readVoidPairs = rddReadPairs.sortByKey(comparator, true, numReducers);
+        } else {
+            readVoidPairs = rddReadPairs.sortByKey(comparator);
+        }
+        return readVoidPairs.keys();
+    }
+
+    /**
+     * Ensure all reads with the same name appear in the same partition.
+     * Requires that the No shuffle is needed.
+
+     */
+    public static JavaRDD<GATKRead> putReadsWithTheSameNameInTheSamePartition(final SAMFileHeader header, final JavaRDD<GATKRead> reads, final JavaSparkContext ctx) {
+        Utils.validateArg(ReadUtils.isReadNameGroupedBam(header), () -> "Reads must be queryname grouped or sorted. " +
+                "Actual sort:" + header.getSortOrder() + "  Actual grouping:" +header.getGroupOrder());
+        int numPartitions = reads.getNumPartitions();
+        final String firstGroupInBam = reads.first().getName();
+        // Find the first group in each partition
+        List<List<GATKRead>> firstReadNamesInEachPartition = reads
```

Comment:
rename to `firstReadNameGroupInEachPartition`

Reply:
done